perfkitbenchmarker/analyze.py

```diff
@@ -0,0 +1,202 @@
+
+  return (start_times >= interval_start) & (record_ends <= interval_end)
+
+
+def AllStreamsThroughputStats(durations, sizes, stream_ids,
+                              num_streams, interval_duration):
+  """Compute the net throughput of multiple streams doing operations.
+
+  Args:
+    durations: a pd.Series of durations in seconds, as floats.
+    sizes: a pd.Series of bytes.
+    stream_ids: a pd.Series of any type.
+    num_streams: int. The number of streams.
+    interval_duration: a float. The time all streams were active, in seconds.
+
+  durations, sizes, and stream_ids must have matching indices. This
+  function computes the per-stream net throughput (sum of bytes
+  transferred / total transfer time) and then adds the results to find
+  the overall net throughput. Operations from the same stream cannot
+  overlap, but operations from different streams may overlap.
+
+  Returns: a tuple of
+    - The net throughput of all streams, excluding times they were inactive.
+    - The throughput of all streams, including times they were inactive.
+    - The total time that all streams were inactive.
+    - The total time all streams were inactive, as a proportion of the
+      total benchmark time.
+  """
+
+  total_bytes_by_stream = sizes.groupby(stream_ids).sum()
+  active_time_by_stream = durations.groupby(stream_ids).sum()
+  total_overhead = interval_duration * num_streams - active_time_by_stream.sum()
+
+  return ((total_bytes_by_stream / active_time_by_stream).sum(),
+          total_bytes_by_stream.sum() / interval_duration,
+          total_overhead,
+          total_overhead / (interval_duration * num_streams))
+
+
+def SummaryStats(series, name_prefix=''):
+
+  """Compute some summary statistics for a series.
+
+  Args:
+    series: a pd.Series of floats.
+    name_prefix: if given, a prefix for the statistic names.
+
+  Returns: a pd.Series with summary statistics.
+  """
+
+  # Percentiles 0, 1, 2, ..., 100
+  percentiles = range(0, 101, 1)
+  # range() and xrange() don't accept floating-point arguments, so
+  # add 99.1, 99.2, ..., 99.9 ourselves.
+  for i in xrange(1, 10):
+    percentiles.append(99 + i / 10.0)
+
+  # TODO: use series.describe() to simplify this once we upgrade to a
+  # version of pandas where describe() has the 'percentiles' keyword
+  # argument.
+  result = {}
+  values = sorted(series)
+  count = len(values)
+
+  for percentile in percentiles:
+    name = name_prefix + 'p' + str(percentile)
+    val = values[int((count - 1) * float(percentile) / 100.0)]
+    result[name] = val
+
+  result[name_prefix + 'min'] = values[0]
+  result[name_prefix + 'max'] = values[-1]
+  result[name_prefix + 'median'] = result[name_prefix + 'p50']
+  result[name_prefix + 'mean'] = sum(values) / float(count)
+
+  sum_of_squares = sum([num ** 2 for num in values])
+  result[name_prefix + 'stddev'] = (sum_of_squares / (count - 1)) ** 0.5
+
+  return pd.Series(result)
+
+
+def AppendStatsAsSamples(series, unit, samples_list,
```

Comment:
I'd like to avoid introducing another function to compute percentiles and append to samples if possible. that's a duplicate to what we already have. See my comments below in the place where this function is called.